sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala

```diff
@@ -159,6 +164,174 @@ object QueryCompilationErrors {
+  }
+
+  def groupingMustWithGroupingSetsOrCubeOrRollupError(): Throwable = {
+    new AnalysisException("grouping()/grouping_id() can only be used with GroupingSets/Cube/Rollup")
+  }
+
+  def pandasUDFAggregateNotSupportedInPivotError(): Throwable = {
+    new AnalysisException("Pandas UDF aggregate expressions are currently not supported in pivot.")
+  }
+
+  def aggregateExpressionRequiredForPivotError(sql: String): Throwable = {
+    new AnalysisException(s"Aggregate expression required for pivot, but '$sql' " +
+      "did not appear in any aggregate function.")
+  }
+
+  def expectedTableNotTempViewError(quoted: String, cmd: String, t: TreeNode[_]): Throwable = {
+    new AnalysisException(s"$quoted is a temp view. '$cmd' expects a table",
+      t.origin.line, t.origin.startPosition)
+  }
+
+  def expectedTableOrPermanentViewNotTempViewError(
+      quoted: String, cmd: String, t: TreeNode[_]): Throwable = {
+    new AnalysisException(s"$quoted is a temp view. '$cmd' expects a table or permanent view.",
+      t.origin.line, t.origin.startPosition)
+  }
+
+  def viewDepthExceedsMaxResolutionDepthError(
+      identifier: TableIdentifier, maxNestedViewDepth: Int, t: TreeNode[_]): Throwable = {
+    new AnalysisException(s"The depth of view $identifier exceeds the maximum " +
+      s"view resolution depth ($maxNestedViewDepth). Analysis is aborted to " +
+      s"avoid errors. Increase the value of ${SQLConf.MAX_NESTED_VIEW_DEPTH.key} to work " +
+      "around this.", t.origin.line, t.origin.startPosition)
+  }
+
+  def insertingIntoViewNotAllowedError(identifier: TableIdentifier, t: TreeNode[_]): Throwable = {
+    new AnalysisException(s"Inserting into a view is not allowed. View: $identifier.",
+      t.origin.line, t.origin.startPosition)
+  }
+
+  def writingIntoViewNotAllowedError(identifier: TableIdentifier, t: TreeNode[_]): Throwable = {
+    new AnalysisException(s"Writing into a view is not allowed. View: $identifier.",
+      t.origin.line, t.origin.startPosition)
+  }
+
+  def writingIntoV1TableNotAllowedError(identifier: TableIdentifier, t: TreeNode[_]): Throwable = {
+    new AnalysisException(s"Cannot write into v1 table: $identifier.",
+      t.origin.line, t.origin.startPosition)
+  }
+
+  def expectsTableNotViewError(v: ResolvedView, cmd: String, t: TreeNode[_]): Throwable = {
+    val viewStr = if (v.isTemp) "temp view" else "view"
+    new AnalysisException(s"${v.identifier.quoted} is a $viewStr. '$cmd' expects a table.",
+      t.origin.line, t.origin.startPosition)
+  }
+
+  def starNotAllowedWhenGroupByOrdinalPositionUsedError(): Throwable = {
+    new AnalysisException(
+      "Star (*) is not allowed in select list when GROUP BY ordinal position is used")
+  }
+
+  def starInExplodeOrJsonTupleOrUDTFInvalidError(): Throwable = {
+    new AnalysisException("Invalid usage of '*' in explode/json_tuple/UDTF")
+  }
+
+  def starInExpressionInvalidError(prettyName: String): Throwable = {
```

Comment:
I think this one can be combined with the previous one (starInExplodeOrJsonTupleOrUDTFInvalidError) into an `invalidStarUsageError` with detailed messages.

Reply:
I think this work should be put in the ticket for error message optimization. @gatorsmile has already established this ticket.

Reply:
I don't have a strong opinion about when to do this. If we plan to do it later, @beliefer can you create a JIRA ticket to track it since it's discovered? I'm also OK if you fix it in this PR.

Reply:
OK